src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1318,29 +1469,29 @@ public static TSource Single<TSource>(this IEnumerable<TSource> source)
 
         public static TSource Single<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            if (predicate == null) throw Error.ArgumentNull("predicate");
-            TSource result = default(TSource);
-            long count = 0;
-            foreach (TSource element in source)
+            Error.CheckSourceNotNull(source);
+            Error.CheckPredicateNotNull(predicate);
+            using (IEnumerator<TSource> e = source.GetEnumerator())
             {
-                if (predicate(element))
+                while (e.MoveNext())
                 {
-                    result = element;
-                    checked { count++; }
+                    TSource result = e.Current;
+                    if (predicate(result))
+                    {
+                        while (e.MoveNext())
+                        {
+                            if (predicate(e.Current)) throw Error.MoreThanOneMatch();
+                        }
+                        return result;
+                    }
                 }
             }
-            switch (count)
-            {
-                case 0: throw Error.NoMatch();
-                case 1: return result;
-            }
-            throw Error.MoreThanOneMatch();
+            throw Error.NoMatch();
```

Comment:
Yes. I think the best course of action is to separate pure optimizations (not observable) into a separate PR and include time/allocation measurements for various scenarios to see that the performance is indeed better.
As for observable changes - some changes like the early-out in Single seems to be good ideas, but it would be easier if they are separate changes. Changes in behavior may need a broader discussion and implications of one could differ from another. It would be easier if those are separately reviewable changes.